### DISCUSSION TITLE
How to implement the parser

### DISCUSSION THREAD

**Discussion by abdukhashimov** [abdukhashimov/student_aggregator#42] [Ideas]
# Parser
 - RS school
 - We are community
 
 For a start two different source are gonna be provided, and their data with xlsx file. How can we parse and store it to database. 
 In all data, the user info will be provided

**Comment by ihoro-epam** [comment]
I have few thoughts on this question:
1. File/other data source parser and Students linker logic should be separated.
2. Are we expecting that user will provide two files ("rs school" data and "We are community" data)?
3. Following all of the above we need to define a complete user story. My proposal is:
```
 1. User uploads data source file(s). System stores the file to blob storage and returns uploaded file ids.
 2. User triggers file parser with the schema to parse. System parses the file, store data to db and returns parsed data (optional)
 3. User triggers linking providing sources refs (file ids, or marking other data sources like 'RSS', 'WAC') and defining primary key to link with (email, family name etc). System trying to link sources and stores linked student to db and returns final student.
```
This will allow us to be more flexible on how many data sources we can use: one, to or more.

We can simplify this flow of-course. But please provide your version.

**Reply by maxwww** [reply]
It works for me.
Yes, for sure we need to separate the logic of loading/parsing/linking. But if the system does not determine the type of scheme maybe we can combine first and second steps together (the user can provide schema during uploading file)? We can combine this on handlers layer, and keep separated logic on services layer.

**Reply by abdukhashimov** [reply]
> 2\. Are we expecting that user will provide two files (

For a starting point we are expecting single file in single request but we are expecting 2 different schemas for initial point.

**Reply by abdukhashimov** [reply]
I have no negative concerns regarding the complete user story. Let's implement it. And break it down to tasks if possible. I now got the idea of parsing from any schema. 
1. user upload the file
2. user provide the sequence of key value pairs 
   [{row: "A", field: "first_name"}, {row: "B", field: "last_name",]
   or just the sequence
   ["first_name", "last_name", score, final_date]
3. specify where to store the parsed information. Either RSS or WAC and our system tries to store the parsed information to the provided index. 
 For example. we can have the defined structure and indexing for RSS and WAC and if the later, new schema needed, there will be just created models for new type of schema. So parser remain the same.

**Reply by abdukhashimov** [reply]
by index I meant the collections :)

**Reply by abdukhashimov** [reply]
Please give objections if I got the idea wrong

**Reply by ihoro-epam** [reply]
> 2. user provide the sequence of key value pairs
>    [{row: "A", field: "first_name"}, {row: "B", field: "last_name",]
>    or just the sequence
>    ["first_name", "last_name", score, final_date]

Im not sure I understood your schema idea.
The file format could have too many schematics (how data is stored in). At this moment we should be more strict on schemas: vertical columns which will represent student fields and rows that will contain individual student data.

**Reply by abdukhashimov** [reply]
For more clarity I will describe detailedly and then create tasks

**Reply by abdukhashimov** [reply]
after couple of hours

**Reply by abdukhashimov** [reply]
*not row, it was column

**Reply by abdukhashimov** [reply]
![image](https://user-images.githubusercontent.com/57385309/201025403-edf89c33-9978-4032-abd6-9b8762338cef.png)

**Reply by abdukhashimov** [reply]
This is going to be the way that we will continue the implementation, any questions and suggestions are welcome!

**Reply by abdukhashimov** [reply]
Feel free to convert parts into tasks, if you are out of tasks @maxwww @ihoro-epam ) Thanks

**Reply by abdukhashimov** [reply]
I will also convert it to tasks asap I have free hands
